### ISSUE TITLE
The Chinese font is garbled and cannot be displayed normally!!!1

### ISSUE THREAD

**Issue by sangko-dgq** [ChordPro/chordpro#497]
![image](https://github.com/user-attachments/assets/cbafe34c-37f3-4af5-9f14-79cbf7362eae)

**Comment by sciurius** [comment]
Please see https://www.chordpro.org/chordpro/faq-unicode-characters/ .

**Comment by zhouhao27** [comment]
Can you provide an example configuration file? 

When I created a custom configuration like this:
```
// Font for typesetting titles.
pdf.fonts.title : "Xingkai SC 14"
```

Restart the app, then it shows the following error message:

> 03:01:35 PM: Warning: Unknown font property: sc at /Applications/ChordPro.app/Contents/MacOS/script/../lib/ChordPro/Output/PDF/Writer.pm line 782.
03:01:35 PM: Warning: No font found for "title"
03:01:35 PM: Error: Unhandled fonts detected -- aborted

**Comment by sciurius** [comment]
I assume your config contains something similar to:
````
// Font for typesetting titles.
pdf.fonts.title : "Xingkai SC 14"

// Mapping font family names to font files.
pdf.fontconfig {
    "Xingkai SC" {
	"": ...actual font file name...
    }
}
````

ChordPro expects a font family name, an optional style (or styles) and an optional size. The styles it recognizes are `bold` and `italic`. "SC" is not a recognized syle, and this causes the error you see.

You can circumvent this by placing an underscore between "Xinkai" and "SC":
````
// Font for typesetting titles.
pdf.fonts.title : "Xingkai_SC 14"

// Mapping font family names to font files.
pdf.fontconfig {
    "Xingkai_SC" {
	"": ...actual font file name...
    }
}
````
Does this help?

**Comment by zhouhao27** [comment]
Thanks for your reply. Here is my new config:

```
pdf.fonts.title : "Xingkai_SC 14"

pdf.fontconfig {
  "Xingkai_SC" {
    "": STXingkaiSC-Light
  }
}
```

But still the same error.

**Comment by sciurius** [comment]
You define the family name as `Xingkai_SC` so you should use this name.
Also, you should add other variants if you need them, e.g.

````
pdf.fontconfig {
  "Xingkai_SC" {
    "": STXingkaiSC-Light
    bold: STXingkaiSC-Bold
  }
}
````

But `STXingkaiSC-Light` is not the name of the **file** that contains the font. You should specify a **complete filename** here.

Then you can try `pdf.fonts.title : "xingkai_sc bold 14"`

If it still fails, please include the full config plus a song, and a screenshot of the error, so we can try to reproduce the problem.

**Comment by zhouhao27** [comment]
@sciurius Still have problem. But this time the error message looks a bit different.

Here is the error message: 

![Image](https://github.com/user-attachments/assets/6cb5a821-23d4-4aa1-b666-40bbb0085728)

Here is my configuration:

![Image](https://github.com/user-attachments/assets/10fbac28-60d2-40de-811c-047692195b80) 

[customconfig.json](https://github.com/user-attachments/files/20051673/customconfig.json)

Here is screen shot for my font file:

![Image](https://github.com/user-attachments/assets/f4a003bf-5574-4885-af7f-d765d6d57f54)

**Comment by sciurius** [comment]
Excellent reporting, thanks. You're almost there.

The one thing that is missing is that in `pdf.fontconfig` you specify the **name** of the `STXingkaiSC-Light` and `STXingkaiSC-Bold` fonts, but you must specify the actual **files** that implement the font.

Most likely you will find the font files in `/System/Library/Fonts` or (in your home folder) `Library/Fonts`.

Currently, ChordPro requires the fonts to be implemented as TrueType (`.ttf`) or OpenType (`.otf`) files. In case the font is implemented as a macOS _font collection_ (`.ttc`) then you must first extract the ttf/otf files from the ttc. There are tools on the internet that can do that for you.

**Comment by zhouhao27** [comment]
> Currently, ChordPro requires the fonts to be implemented as TrueType (`.ttf`) or OpenType (`.otf`) files. In case the font is implemented as a macOS _font collection_ (`.ttc`) then you must first extract the ttf/otf files from the ttc. There are tools on the internet that can do that for you.

It's `.ttc` file. But I could not find where it is.

![Image](https://github.com/user-attachments/assets/c181c70e-cedb-4a54-bf8c-5b8d1830cc94)

**Comment by zhouhao27** [comment]
I still got the same issue when I use `.otf` font.

```
pdf.fonts.title : "LingWai_SC 14"
pdf.fontconfig {
  "Xingkai_SC" {
    "": STXingkaiSC-Light
    bold: STXingkaiSC-Bold
  },
  "LingWai_SC" {
    "": LingWaiSC-Medium
  }
}
```

`LingWaiSC-Medium` is the real file name.

![Image](https://github.com/user-attachments/assets/75816327-78f9-44dc-bd8d-4064bc951bc2)

**Comment by sciurius** [comment]
> `LingWaiSC-Medium` is the real file name.

I don't think so. The filename should be something like `/System/Library/Fonts/LingWaiSC-Medium.otf`.

**Comment by zhouhao27** [comment]
> > `LingWaiSC-Medium` is the real file name.
> 
> I don't think so. The filename should be something like `/System/Library/Fonts/LingWaiSC-Medium.otf`.

Don't get you. 
```
pdf.fonts.title : "sans bold 14"
pdf.fontconfig.sans.bold       : Helvetica-Bold
```

This one doesn't including the file extension. And I can't find `LingWaiSC-Medium.otf` in `/System/Library/Fonts` or `/Library/Fonts` or `~/Library/Fonts`.

**Comment by Desbeers** [comment]
@zhouhao27 

> This one doesn't including the file extension. And I can't find `LingWaiSC-Medium.otf` in `/System/Library/Fonts` or `/Library/Fonts` or `~/Library/Fonts`.

If you open the `Font Book` application on your Mac and *right click* on the font it will show you where it is in the Finder. Hope that helps!

**Comment by Desbeers** [comment]
> No. It doesn't show the path because Show in Finder is disabled. I found the installed fonts are in /System/Library/AssetsV2/com_apple_MobileAsset_Font7. Will this cause the issue?

Yes, I think thats the issue. It’s an additional downloaded font from an Apple server and its restricted in its usage:

<img width="242" alt="Image" src="https://github.com/user-attachments/assets/0e871180-840c-482f-b39e-e9f8e9bbb45c" />

I’ve tried the font myself and for me it does not work either. You have to try it with another font. You can see the optional restrictions in `Font Book`.

**Comment by zhouhao27** [comment]
I changed to use the system font. Still got the same issue.

![Image](https://github.com/user-attachments/assets/fc51d38a-d87f-4c9d-8d60-cd6ce3b0dc88)

My configuration:

```
pdf.fonts.title : "Songti_SC bold 14"
pdf.fontconfig {
  "Songti_SC" {
    "": Songti 
  }
}
```

**Comment by Desbeers** [comment]
The `Songti` font is a font *collection* (ttc font). Those fonts are not supported by ChordPro. Only *ttf* and *otf* is.

> /System/Library/Fonts/Supplemental/Songti.ttc

You have to extract the fonts from the collection if you want to use them. A website such as https://transfonter.org/ttc-unpack can do that.

**Comment by zhouhao27** [comment]
> https://transfonter.org/ttc-unpack

Even I extract out the `ttf` file, I can not install since it's already installed. I can not uninstall system fonts. By the why `Helvetica` can be used? It's also `.ttc` file.

**Comment by Desbeers** [comment]
> Even I extract out the ttf file, I can not install since it's already installed. I can not uninstall system fonts. By the why Helvetica can be used? It's also .ttc file.

You don’t have to install the extracted fonts to use it in ChordPro. Just point your ChordPro config to the **full** file location:

Example:

    "fonts" : {
      "text" : {
        "color" : "#000000",
        "file" : "\/var\/folders\/sy\/08sgh__57lv_h2tf3pk307z40000gp\/T\/ChordProviderTMP\/STSongti-SC-Regular.ttf",
        "size" : 12
      },
    },
    

Helvetica can be used because its a build-in font for ChordPro; its not using the system font by default.

**Comment by sciurius** [comment]
> \/var\/folders\/sy\/08sgh__57l...

Why the backslashes?

**Comment by Desbeers** [comment]
> Why the backslashes?

Ah, my JSON encoder does that. I *autogenerate* a ChordPro config in my Chord Provider application to use with the ChordPro CLI. By quoting a path like this I can use paths that contains spaces.

**Comment by zhouhao27** [comment]
> > Even I extract out the ttf file, I can not install since it's already installed. I can not uninstall system fonts. By the why Helvetica can be used? It's also .ttc file.
> 
> You don’t have to install the extracted fonts to use it in ChordPro. Just point your ChordPro config to the **full** file location:
> 
> Example:
> 
> ```
> "fonts" : {
>   "text" : {
>     "color" : "#000000",
>     "file" : "\/var\/folders\/sy\/08sgh__57lv_h2tf3pk307z40000gp\/T\/ChordProviderTMP\/STSongti-SC-Regular.ttf",
>     "size" : 12
>   },
> },
> ```
> 
> Helvetica can be used because its a build-in font for ChordPro; its not using the system font by default.

Can you share your whole configuration file? Thanks.

**Comment by Desbeers** [comment]
Ah, great!

**Comment by Desbeers** [comment]
So we can close this issue now?
